### PR TITLE
fix(hunyuan-video): align text conditioning and unstick trainside GRPO

### PR DESCRIPTION
**Pull request by leviking98z-rgb**
## Summary

Consolidates the HunyuanVideo-1.0 work from #220 and the now-closed #215:

1. **Use canonical LLaMA conditioning.** The trainside encoder previously used `outputs.last_hidden_state`; official HunyuanVideo, Diffusers, and the SGLang path use the third-from-last state, `hidden_states[-3]`. This adds `hidden_state_skip_layer` (default `2`), threads it through the pipeline, rejects negative/out-of-range values, and keeps `0` available for the legacy baseline.
2. **Update the canonical trainside recipe in place.** The default recipe now uses the validated climbing-PickScore settings: LoRA rank 64 / alpha 256, batch 8, 8 samples per prompt, 10 denoising steps, standard landscape 1280x720x5 output, and 4 SDE steps. It also restores two PPO updates and the complete FSDP block list.
3. **Add the matching SGLang recipe.** `hunyuan_video_t2v_sglang.yaml` uses the same 1280x720x5 training/sampling defaults and a local TP=1 SGLang engine on each of 8 GPUs. It trains LoRA with FSDP and sends merged transformer weights through `TensorWeightSync`, avoiding the known online-SGLang-LoRA delta mismatch.
4. **Make HunyuanVideo a real video rollout adapter.** The adapter now returns packed video frames for `VideoPickScore` and preserves SGLang's two text streams as `text_llama` and `pooled_clip`, including variable-length LLaMA padding and attention masks required by trainer-side replay.
5. **Make 720p execution reliable.** CUDA text encoding disables the unstable cuDNN SDPA backend while retaining PyTorch's other SDPA backends. Both 720p recipes enable Diffusers VAE spatial tiling to bound fp32 decode activation memory.

## Related Issue

N/A. Supersedes the closed #215, whose recipe work is consolidated here.

## Test Plan

- Repository checks:
  `SKIP=no-commit-to-branch pre-commit run --all-files --show-diff-on-failure`
  — all hooks passed, including YAML, Python AST, ruff, private-key detection, and recipe `_target_` resolution.
- NVIDIA H20 CUDA/bf16 hidden-state smoke with a real four-layer `transformers.LlamaModel`: `skip=0` exactly matched `hidden_states[-1]`, and `skip=2` exactly matched `hidden_states[-3]` (zero tolerance).
- Real **8x NVIDIA H20** end-to-end SGLang run with the canonical recipe and only `num_rollouts=1` overridden:
  - all 8 local SGLang engines initialized;
  - generated 64 videos at 1280x720x5 through all 10 denoising steps;
  - completed video packing, trainer-side replay, `VideoPickScore`, merged-weight synchronization, and the training loop;
  - exit code **0**, reward `0.7297`, ratio `1.0000`, and approximately 30.9 GiB peak memory per GPU.
- Real **8x NVIDIA H20** end-to-end trainside run with the canonical recipe and only `+num_rollouts=1` plus W&B disablement overridden:
  - completed 1280x720x5 sampling, tiled VAE decode, trainer-side replay, `VideoPickScore`, and the training loop;
  - exit code **0**, reward `0.7239`, ratio `1.0000`, and approximately 43.5-45.0 GiB observed memory per GPU;
  - the tiled decode passed the point where whole-frame fp32 decode previously attempted a 118.65 GiB allocation.
- Existing controlled training evidence (VideoPickScore, lr `3e-4`, clip `1e-4`):
  - 8 GPUs: alpha 64 stayed flat, alpha 128 was approximately flat, alpha 256 climbed, and alpha 512 climbed but overshot at high learning rate; fp32 with alpha 64 remained flat.
  - 32 GPUs, alpha 256: reward improved from about 0.703 to 0.831 over 150 rollouts.
  - Paired trainside/SGLang validation with skip 2 and alpha 256 tracked within about +/-0.02 through rollout 44, mostly within +/-0.01, with ratio 1.0000 and no collapse.

Reproduce the canonical trainside recipe:

```bash
DATA_PATH=datasets/pickscore/train.txt \
EVAL_DATA_PATH=datasets/pickscore/test.txt \
bash examples/run_experiment_single_node.sh \
  diffusion/hunyuan_video/hunyuan_video_t2v_trainside \
  +num_rollouts=150
```

Reproduce the canonical SGLang recipe:

```bash
DATA_PATH=datasets/pickscore/train.txt \
EVAL_DATA_PATH=datasets/pickscore/test.txt \
bash examples/run_experiment_single_node.sh \
  diffusion/hunyuan_video/hunyuan_video_t2v_sglang
```

## Compatibility / Risk

- This intentionally changes the **default** HunyuanVideo trainside behavior: skip `0 -> 2`, alpha `64 -> 256`, batch `48 -> 8`, samples per prompt `16 -> 8`, inference steps `20 -> 10`, resolution `832x480 -> 1280x720`, and SDE steps `8 -> 4`.
- Existing experiments can reproduce the prior values with Hydra overrides; setting `bundle.config.hidden_state_skip_layer=0` also propagates to the pipeline.
- `vae_use_tiling` is optional and defaults to `false`; the two 720p recipes enable it. No checkpoint or data-format migration is required.
- Disabling cuDNN SDPA changes CUDA attention backend selection process-wide in HunyuanVideo rollout workers; flash, memory-efficient, and math SDPA remain available.
- 1280x720 costs more compute and memory than the previous 832x480 trainside default. The end-to-end H20 runs above cover both execution paths at the new default.
- The SGLang recipe runs one TP=1 full-model engine per GPU and performs merged full-weight synchronization, which costs more GPU memory and synchronization bandwidth than trainside rollout.

## Reviewer Notes

- #215 was closed and folded into this PR. Following the maintainer suggestion there, the original trainside recipe is updated in place; the new file is a distinct SGLang execution path rather than a duplicate trainside variant.
- Duplicate-work check: no other open PR directly targets this HunyuanVideo correction.
- AI assistance was used. The full final diff was self-reviewed, tested after merging the latest `main`, and contains no generated artifacts or credentials.
- Official HunyuanVideo and Diffusers both use hidden-layer skip 2; both UniRL rollout paths now use the same conditioning and 1280x720x5 sampling defaults.
- The SGLang recipe intentionally uses `lora_merged: true`; online LoRA adapter deltas are not used.
- The cuDNN SDPA guard and tiled VAE decode were added from failures reproduced during real H20 end-to-end testing.

Validation plot from the skip-layer experiment:

HunyuanVideo skip-layer validation

## Checklist

- [x] I reviewed the changed code and removed unrelated/generated artifacts.
- [x] I updated docs and configs where needed; standalone PR tests were intentionally removed.

### CODE REVIEW

**Comment by leviking98z-rgb** [comment]
## Validation: paired alignment on the corrected (skip=2) config — confirmed

Ran trainside(skip=2) and the sglang rollout(skip=2) as a paired curve at matched settings (16 GPU each; 480×640, 10 steps, num_sde_steps=4, spp=8, batch=16, α=256, video_pickscore). Both engines now condition on the canonical `hidden_states[-3]`.

Matched-rollout reward (both `ratio=1.0000`, `clip=0`, no collapse):

| rollout | trainside | sglang | gap |
|---|---|---|---|
| r1  | 0.7106 | 0.7043 | +0.006 |
| r30 | 0.7965 | 0.7948 | +0.002 |
| r35 | 0.7941 | 0.7942 | −0.000 |
| r40 | 0.7981 | 0.7987 | −0.001 |
| r42 | 0.7941 | 0.8141 | −0.020 |
| r44 | 0.7695 | 0.7677 | +0.002 |

Both cross r40 tracking within **±0.02** (mostly ≤±0.01) — the two engines produce the same curve on the corrected config.

Also note both start/track **higher** than the legacy skip=0 baseline (r1 ≈ 0.71 vs ≈ 0.685; the old o2c8gg0k skip=0 curve peaked ~0.83 at r150), consistent with skip=2 feeding the DiT the layer it was distilled for → better prompt adherence.

Marking ready for review.

**Review by KemingWu — APPROVED** [review]
LGTM
